internal/daemon/controller/handlers/targets/target_service.go

```diff
@@ -766,25 +770,32 @@ func (s Service) AuthorizeSession(ctx context.Context, req *pbs.AuthorizeSession
 		}
 	}
 
-	if chosenEndpoint == nil {
-		if len(endpoints) == 0 {
-			// No hosts were found, error
-			return nil, handlers.NotFoundErrorf("No endpoint found from available target host sources.")
-		}
+	var h, p, hostId, hostSetId string
+	if chosenEndpoint == nil && len(endpoints) > 0 {
 		chosenEndpoint = endpoints[rand.Intn(len(endpoints))]
 	}
 
-	h, p, err := net.SplitHostPort(chosenEndpoint.Address)
-	switch {
-	case err != nil && strings.Contains(err.Error(), missingPortErrStr):
-		if t.GetDefaultPort() == 0 {
-			return nil, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("neither the selected host %q nor the target provides a port to use", chosenEndpoint.HostId))
+	if chosenEndpoint != nil {
+		hostId = chosenEndpoint.HostId
+		hostSetId = chosenEndpoint.SetId
+		h, p, err = net.SplitHostPort(chosenEndpoint.Address)
+		switch {
+		case err != nil && strings.Contains(err.Error(), missingPortErrStr):
+			if t.GetDefaultPort() == 0 {
+				return nil, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("neither the selected host %q nor the target provides a port to use", chosenEndpoint.HostId))
```

Comment:
Could this error be encountered by a user? If so, maybe we should use `handlers.ApiErrorWithCodeAndMessage`?

Reply:
This error message was not written by me, I only reorganized the code to better handle when we need to parse from a chosen host resource or a network address that is directly associated to a target. I'm happy to change it though.